File: leetcode/222-count-complete-tree-nodes/main.go
package main

import "math"

// Definition for a binary tree node.
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

/*
	1st approach dfs
	Time		O(n)
	Space		O(height of the tree)
	24ms beats 92.86%
	20jan2019
*/
func countNodes(root *TreeNode) int {
	count := 0
	var dfs func(node *TreeNode)
	dfs = func(node *TreeNode) {
		if node == nil {
			return
		}
		count++
		dfs(node.Left)
		dfs(node.Right)
	}
	dfs(root)
	return count
}

/*
	2nd approach
	- actually we dont need to go though all the nodes
	- we can just search for the node which the left and right have different height
	Time		O(h*h) worst
	Space		O(height of the tree)
	20ms beats 100%
	20jan2019
*/
func countNodes1(root *TreeNode) int {
	if root == nil {
		return 0
	}
	left := 1 + getLeftHeight(root)
	right := 1 + getRightHeight(root)
	if left == right {
		// if left == right, it means that this subtree is a perfect tree
		// therefore the number of nodes = 2^n-1
		return int(math.Pow(2, float64(left))) - 1
	} else {
		// if unfortunately left != right, we need traverse down its children and count
		return 1 + countNodes1(root.Left) + countNodes1(root.Right)
	}
}

func getLeftHeight(root *TreeNode) int {
	if root == nil {
		return 0
	}
	height := 0
	for root.Left != nil {
		height++
		root = root.Left
	}
	return height
}

func getRightHeight(root *TreeNode) int {
	if root == nil {
		return 0
	}
	height := 0
	for root.Right != nil {
		height++
		root = root.Right
	}
	return height
}

func main() {

}

File: leetcode/222-count-complete-tree-nodes/main.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

"""
    2nd approach
	- actually we dont need to go though all the nodes
	- we can just search for the node which the left and right have different heights
        - if no diff, the subtree has 2^n-1 nodes
        - if has diff, go into its children and count the number of nodes again

    Time    O(logn * logn) because it is almost balenced
    Space   O(h)
    80 ms, faster than 80.30%
    24apr2019
"""


class Solution(object):
    def countNodes(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        if root == None:
            return 0
        left = 1 + self.countLeftMost(root.left)
        right = 1 + self.countRightMost(root.right)
        if left == right:
            # if left == right, it means that this subtree is a perfect tree
            # therefore the number of nodes = 2^n-1
            return 2**left-1
        # if unfortunately left != right, we need to traverse down its children and count
        return self.countNodes(root.left) + self.countNodes(root.right) + 1

    def countLeftMost(self, node):
        if node == None:
            return 0
        count = 0
        cur = node
        while cur != None:
            count += 1
            cur = cur.left
        return count

    def countRightMost(self, node):
        if node == None:
            return 0
        count = 0
        cur = node
        while cur != None:
            count += 1
            cur = cur.right
        return count
